File: python/reverse_nodes_in_k_group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            kth = self.getKth(groupPrev, k)
            if not kth:
                break
            
            groupNext = kth.next
            prev, cur = groupNext, groupPrev.next

            while cur != groupNext:
                nxt = cur.next
                cur.next = prev
                prev = cur
                cur = nxt
            temp = groupPrev.next
            groupPrev.next = prev
            groupPrev = temp
        return dummy.next

    def getKth(self, cur, k):
        temp = cur
        while temp:
            if not k:
                return temp
            k -= 1
            temp = temp.next
        return None

File: C++/reverse_nodes_in_k_group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* groupPrev = dummy;

        while (true){
            ListNode* kth = getKth(groupPrev, k);
            if (kth == nullptr){
                break;
            }

            ListNode* groupNext = kth->next;
            ListNode* prev = groupNext; ListNode* cur = groupPrev->next;

            while (cur != groupNext){
                ListNode* nxt = cur->next;
                cur->next = prev;
                prev = cur;
                cur = nxt;
            }

            ListNode* temp = groupPrev->next;
            groupPrev->next = kth;
            groupPrev = temp;
        }

        return dummy->next;
    }

    ListNode* getKth(ListNode* cur, int k){
        ListNode* temp = cur;
        while (temp != nullptr){
            if (k == 0){
                return temp;
            }
            k--;
            temp = temp->next;
        }
        return nullptr;
    }
};
